src/device-state/db-format.ts

```diff
@@ -0,0 +1,185 @@
+// Currently this function does quite a bit more than it needs to as it pulls in a bunch
+// of required information for the instances but we should think about a method of not
+// requiring that data here
+export async function getApp(id: number): Promise<InstancedApp> {
+	const dbApp = await getDBEntry(id);
+	return await buildApp(dbApp);
+}
+
+export async function getApps(): Promise<InstancedAppState> {
+	const dbApps = await getDBEntry();
+	const apps: InstancedAppState = {};
+	for (const app of dbApps) {
+		apps[app.appId] = await buildApp(app);
+	}
+	return apps;
+}
+
+async function buildApp(dbApp: targetStateCache.DatabaseApp) {
+	const volumes = _.mapValues(JSON.parse(dbApp.volumes) ?? {}, (conf, name) => {
+		if (conf == null) {
+			conf = {};
+		}
+		if (conf.labels == null) {
+			conf.labels = {};
+		}
+		return Volume.fromComposeObject(name, dbApp.appId, conf);
+	});
+
+	const networks = _.mapValues(
+		JSON.parse(dbApp.networks) ?? {},
+		(conf, name) => {
+			if (conf == null) {
+				conf = {};
+			}
+			return Network.fromComposeObject(name, dbApp.appId, conf);
+		},
+	);
+
+	const opts = await config.get('extendedEnvOptions');
+	const supervisorApiHost = dockerUtils
+		.getNetworkGateway(constants.supervisorNetworkInterface)
+		.catch(() => '127.0.0.1');
+	const hostPathExists = {
+		firmware: await pathExistsOnHost('/lib/firmware'),
+		modules: await pathExistsOnHost('/lib/modules'),
+	};
+	const hostnameOnHost = _.trim(
+		await fs.readFile(
+			path.join(constants.rootMountPoint, '/etc/hostname'),
+			'utf8',
+		),
+	);
+
+	const svcOpts = {
+		appName: dbApp.name,
+		supervisorApiHost,
+		hostPathExists,
+		hostnameOnHost,
+		...opts,
+	};
+
+	// In the db, the services are an array, but here we switch them to an
+	// object so that they are consistent
+	const services = _.keyBy(
+		await Promise.all(
+			(JSON.parse(dbApp.services) ?? []).map(
+				async (svc: ServiceComposeConfig) => {
+					// Try to fill the image id if the image is downloaded
+					let imageInfo: ImageInspectInfo | undefined;
+					try {
+						imageInfo = await images.inspectByName(svc.image);
+					} catch (e) {
+						if (NotFoundError(e)) {
+							imageInfo = undefined;
+						} else {
+							throw e;
+						}
+					}
+
+					const thisSvcOpts = {
+						...svcOpts,
+						imageInfo,
+						serviceName: svc.serviceName,
+					};
+					// We force the casting here as we know that the UUID exists, but the typings do
+					// not
+					return Service.fromComposeObject(
+						svc,
+						(thisSvcOpts as unknown) as DeviceMetadata,
+					);
+				},
+			),
+		),
+		'serviceId',
+	) as Dictionary<Service>;
+
+	return {
+		appId: dbApp.appId,
+		commit: dbApp.commit,
+		releaseId: dbApp.releaseId,
+		name: dbApp.name,
+		source: dbApp.source,
+
+		services,
+		volumes,
+		networks,
+	};
+}
+
+export async function setApps(
+	apps: { [appId: number]: TargetApplication },
+	source: string,
+	trx?: db.Transaction,
+) {
+	const cloned = _.cloneDeep(apps);
```

Comment:
This can be avoided, as far as I can see the only possible mutation of `apps` this function might do is in `await images.normalise(s.image)` - unless you're worried about the caller mutating it during the async parts of this function (which is possible but I would expect unlikely)